proto/external_storage.proto

```diff
@@ -0,0 +1,72 @@
+syntax = "proto3";
+// purposely leaving out delete. that is up to individual external
+// storage server operators to do on their end
+
+service ProxyStorage {
+
+    rpc GetModuleModfile (ModuleMetadata) returns (ModuleMod) {}
+    rpc GetModuleInfo (ModuleMetadata) returns (ModuleInfo) {}
+    rpc GetModuleZip (ModuleMetadata) returns (ModuleZip) {}
+    rpc SaveModule (ModuleData) returns (ModuleSaveResponse) {}
+    rpc ListVersions (ListVersionsRequest) returns (ListVersionsResponse) {}
+    rpc CheckModuleAndVersion(ModuleMetadata) returns (CheckModuleAndVersionResponse) {}
+    rpc GetModuleCatalog (GetModuleCatalogRequest) returns (GetModuleCatalogResponse) {}
+}
+
+// ModuleMetata is the name and version of a module
+message ModuleMetadata {
+    string name = 1;
+    // this is redundant for GetModuleCatalogResponse but not for 
+    // GetModule and others. Implementations can return this and Athens
+    // may choose to ignore it.
+    string version = 2;
+}
+
+// ModuleMod is the go.mod file for a module@version
+message ModuleMod {
+    string mod_file = 1;
+}
+
+// ModuleInfo is the JSON metadata that gets returned from
+// the @v/v1.2.3.info endpoint
+message ModuleInfo {
+    string timestamp = 1;
+    string version = 2;
+}
+
+// ModuleZip is the zip file of a module's source code
+message ModuleZip {
+    bytes zip_file = 1;
```

Comment:
Wouldn't this be a big performance problem? Go by default loads 10 dependencies concurrently as far as I recall, and if you have multiple go clients, this can quickly become a problem. 

Would be interesting to test it against a couple of concurrent `go mod download`s and see how it performs.

Reply:
@marwan-at-work if the problem you're talking about is how many concurrent calls to `GetZip` Athens is making, I think we can use `GoGetWorkers` and the accompanying machinery to limit the number of times we hit that endpoint. The other side could also decide how many concurrent calls that it wants to serve concurrently.

I don't think we get any worse from a memory usage perspective if we limit concurrent calls to the `GetZip` like we do with `go mod download`. Streaming definitely doesn't hurt us though.

Reply:
> if the problem you're talking about is how many concurrent calls to GetZip Athens is making

That's just one of the problems :) I think the biggest one is the fact that we'd have to completely load the module zip in memory no its way in and on its way out. Even if we limit save/get operations, then the Go command will time out because it would simply take too long to try and process less than 10 gets at a time.